# What is Fuzzy Matching
Fuzzy matching is a technique used to compare and match two strings that are not exactly identical. While the naive approach of comparing strings and checking for equality would only make a direct comparison that is binary and doesn't consider the degree to which two strings might match each other. Fuzzy matching algorithms use various techniques to measure the similarity between two strings, even if they're not exactly identical. These techniques can include things like measuring the number of characters or words that are the same, or looking for common patterns or sequences of characters. By taking a more nuanced approach to comparing strings, fuzzy matching can help identify matches even when there are slight differences or variations in the data being compared.

In [18]:
from thefuzz import fuzz # python lib for fuzzy logic

string1 = "Hello World"
string2 = "hello world"

print(string1 == string2)  # True or False equality
print(f'The strings have a {fuzz.partial_ratio(string1,string2)}% similarity') # Degree to which we belive the strings to be matching

False
The strings have a 82% similarity


The codeblock above shows the value of using Fuzzy Matching. We could have potentially lost out on a word/phrase that has a signifcantly high similarity.

# Use cases of Fuzzy Matching
- Clean Customer Data: Companies use fuzzy matching to find and merge duplicate customer records. This helps to keep the customer database accurate and up-to-date, which is important for sales and marketing efforts.

- Match Medical Records: Healthcare providers use fuzzy matching to match patient records across different hospitals or clinics. This helps to ensure that patients receive consistent and accurate medical care.

- Detect Fraud: Fuzzy matching can help to identify suspicious patterns in financial data, such as transactions that are similar but not identical. This can be used to detect potential fraud.

- Personalize Product Recommendations: E-commerce websites use fuzzy matching to recommend products to users based on their search history or past purchases. This can help to improve the user experience and increase sales.

- Verify Addresses: Postal services and shipping companies use fuzzy matching to verify and correct address information. This helps to ensure that packages are delivered to the correct address and reduces the risk of misdeliveries or lost packages.

# Approaches to computing similarity of two Strings
Lets consider two strings String 1 and String 2

- Number of edits required to transition from String 1 to String 2
    - Edits would consist of the number of character replacements, deletions or insertions.
- Common word counts/tallies
- Frequency of letters and phrases
- Longest common Substrings

## Overview of core Fuzzy Matching Algorithms 

### Levenshtein distance
The Levenshtein distance algorithm is a way to compare two words or strings to see how different they are from each other. It counts how many changes you need to make to one word to turn it into the other. The changes can be deleting a letter, adding a letter, or changing a letter. The fewer changes you need to make, the more similar the words are. 

### Soundex and Metaphone
Soundex is an algorithm that is used to index words by their sound, so that similar-sounding words can be matched even if they are spelled differently. It works by taking a word, dropping its vowels (except for the first letter), and assigning a code to each of the remaining consonants based on their sound. Consonants that have the same sound get the same code. The resulting code can then be used to compare words and find matches. 

Metaphone is an algorithm that indexes words by their pronunciation, just like Soundex. It's better than Soundex because it can handle more complex phonetic patterns and a wider range of words from different languages. It works by assigning codes to the sounds in a word based on their phonetic properties and the context of the sounds. The resulting codes are shorter and easier to read than Soundex codes.

In [19]:
import jellyfish

name1= "Phillip"
name2= "Fillipe"

soundexCode1 = jellyfish.soundex(name1)
soundexCode2 = jellyfish.soundex(name2)
metCode1 = jellyfish.metaphone(name1)
metCode2 = jellyfish.metaphone(name2)
print(soundexCode1)
print(soundexCode2)
print(metCode1)
print(metCode2)
print(f'{fuzz.partial_ratio(name1, name2)}% similarity')


P410
F410
FLP
FLP
71% similarity


The above codeblock shows that the Soundex algorithm produces similar codes showing similar sounding words, but metaphone shows that they sound the exact same. Levenshtein distance doesnt produce a satifatory result with just 71% similarity.

### Cosine Similarity 
For cosine similarity between two terms we make use of the following identity 

cos(θ) = X ⋅ Y / |X| . |Y|

We have two key values to consider cos(90) and cos(0). If the value of thetha is 90 that results in our identity producing 0 showing no similarity. if the value of theta is 0 that results in our identity producing a value of 1. We can obtain the vector of our words using the Bag of words encoding technique.
